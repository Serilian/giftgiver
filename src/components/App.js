import React, {Component} from 'react';
import {Button} from "react-bootstrap";
import Gift from './Gift';
import {maxNumber} from "../helper";

class App extends Component {

    constructor(props) {
        super(props);
        this.state = {
            gifts: []
        };
    }

    addGift = () => {
        const {gifts} = this.state;
        const ids = gifts.map(gift=> gift.id);
        gifts.push({id: maxNumber(ids) + 1});
        this.setState({gifts});
    };

    removeGift = (id)=> {
        const updatedGifts = this.state.gifts.filter(gift => gift.id !== id);
        this.setState({gifts: updatedGifts});
    };

    render() {
        return (
            <div>
                <h2>Gift Giver</h2>
                <div className={"gift-list"}>{this.state.gifts.map(gift=>(
                    <Gift
                        gift={gift}
                        removeGift={this.removeGift}
                        key={gift.id}>Gift</Gift>
                ))}</div>
                <Button variant='primary' className='btn-add' onClick={this.addGift}>Add gift</Button>
            </div>
        );
    }
}

export default App;