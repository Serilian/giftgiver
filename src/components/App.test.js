import React from 'react';
import {shallow} from 'enzyme';
import App from './App.js';

describe('app component', () => {
    let app;
    const MOCK_INDEX = 1;

    beforeEach(() => {
            app = shallow(<App/>);
        }
    );

    it('renders correctly', () => {
        expect(app.debug()).toMatchSnapshot();
    });

    it('initializes the `state` with empty list of gifts', () => {
        expect(app.state().gifts).toEqual([]);
    });


    describe('When clicking add gift button', () => {

        beforeEach(()=>{
            app.find('.btn-add').simulate('click');
        });
        afterEach(()=>{
            app.setState({gifts: []})
        });

        it('then adds new gift to `state`', () => {
            expect(app.state().gifts).toEqual([{id: MOCK_INDEX}]);
        });

        it("then adds new div to rendered list", () => {
            expect(app.find(".gift-list").children().length).toEqual(MOCK_INDEX);
        });

        it('creates a gift component', ()=>{
            expect(app.find('Gift').exists()).toBeTruthy();
        });
    });

    describe('When user wants to remove added gift', () => {
        beforeEach(()=>{
           app.instance().removeGift(MOCK_INDEX);
        });

        it("removes the gift fom state", ()=>{
            expect(app.state().gifts).toEqual([]);
        });

    })


});


