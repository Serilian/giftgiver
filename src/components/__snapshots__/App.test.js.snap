// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app component renders correctly 1`] = `
"<div>
  <h2>
    Gift Giver
  </h2>
  <div className=\\"gift-list\\" />
  <Button variant=\\"primary\\" className=\\"btn-add\\" onClick={[Function]} active={false} disabled={false} type=\\"button\\">
    Add gift
  </Button>
</div>"
`;
